### PR TITLE
Datetime display options

### LINKED ISSUE #428: Implement display options for date & time types
## Problem
The Date & Time Mathesar type supports the following display options:
- Date format
- Time format

We will combine this into one `format` store.

## Solution
(1) We should store these display options in the following format in the `display_options` field of the corresponding column.
```
{
  "format": "YYYY-MM-DD hh:mm A",
}
```

(2) We should also validate these so that:
- Only columns of this type can have these display options. They should not be able to be set if the column is of a different type.
- `format` should be a valid JavaScript date/time format.
- We should validate format based on the type of the column. `DATE` columns should only include date formats, `TIME` column should only include time formats, `TIMESTAMP` columns can have either or both.

(3) If the column type is changed, the display options should be deleted.

(4) We should add supported display options to the `types` endpoint.

## Additional Context
- [Design of Date/Time type options on Figma](https://www.figma.com/proto/Uaf1ntcldzK2U41Jhw6vS2/Mathesar-MVP?page-id=4260%3A37440&node-id=4270%3A41050&viewport=324%2C48%2C0.29&scaling=contain&starting-point-node-id=4270%3A41050&show-proto-sidebar=1)
- Blocked by #658
- #392

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #428

Accept display options for datetime column types. 

**Technical details**
This PR adds a serializer to accept and validate datetime column `display_options`. It uses [arrow library](https://github.com/arrow-py/arrow) to validate `format` field values to accept only valid datetime object and restrict the format string based on types(date column can't accept format string with time formats).

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by silentninja** [comment]
Blocked by #865

File: mathesar/api/serializers/shared_serializers.py [hunk 1]
```diff
@@ -106,9 +108,106 @@ class NumberDisplayOptionSerializer(OverrideRootPartialMixin, serializers.Serial
...
+        if 'z' in display_format.lower():
+            raise serializers.ValidationError(
+                "Timestamp without timezone column cannot contain timezone display format")
+
+
+class DateFormatValidator(AbstractDateTimeFormatValidator):
+
+    def validate(self, datetime_obj, display_format):
+        date_obj = arrow.get('2013-09-30')
+        if datetime_obj.time() != date_obj:
```
Comment (mathemancer) [inline mathesar/api/serializers/shared_serializers.py:153]: Is this correct?  Shouldn't it be
```suggestion
        if datetime_obj.time() != date_obj.time():
```

Maybe I'm missing something.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Overall, this looks pretty good.  Some comments:
- The validation is spread over two functions called in different ways, which is a bit confusing.  Couldn't one call the other?
- One of the validations was missing a piece (I think; see the specific comment)
- Please make `vulture` happy so the pipeline passes.

File: mathesar/api/serializers/shared_serializers.py [hunk 2]
```diff
@@ -106,9 +108,106 @@ class NumberDisplayOptionSerializer(OverrideRootPartialMixin, serializers.Serial
...
+            timestamp_with_tz_obj = arrow.get('2013-09-30T15:34:00.000-07:00')
+            parsed_datetime_str = timestamp_with_tz_obj.format(value)
+            datetime_object = arrow.get(parsed_datetime_str, value)
+        except ValueError:
+            raise serializers.ValidationError(f"{value} is not a valid format used for parsing a datetime.")
+        else:
+            self.validate(datetime_object, value)
+
+    @abstractmethod
+    def validate(self, datetime_obj, display_format):
```
Comment (silentninja) [inline mathesar/api/serializers/shared_serializers.py:131]: > The validation is spread over two functions called in different ways, which is a bit confusing. Couldn't one call the other?

@mathemancer Do you mean this validation function?

Comment (mathemancer) [reply mathesar/api/serializers/shared_serializers.py:131]: Whoops.  I completely missed the `else`.  My fault.  disregard that.

File: mathesar/api/serializers/shared_serializers.py (hunk 1 above)
Comment (silentninja) [reply mathesar/api/serializers/shared_serializers.py]: Yes, it should be checking the time part of the object. I have made the necessary changes, thanks!

**Review by mathemancer — APPROVED** [review]
Looks great to me; thanks!

**Pull request merged**